mcs fails to compile ParseNumber.cs

<!--
If you are new to the project get yourself familiar with https://www.mono-project.com/community/bugs/make-a-good-bug-report/ before filling the issue
-->

## Steps to Reproduce
The issue was found in https://github.com/mono/mono/pull/8626 
MCS crashes on the following code (simplified ParseNumbers.cs from corefx, line 515):
```csharp
namespace ConsoleApplication16
{
    internal class Program
    {
        public static void Main()
        {
            while (true && IsDigit(out int value))
                value++;
        }
        
        static bool IsDigit(out int p)
        {
            p = 0;
            return true;
        }
    }
}
```
It looks like the issue is somewhere in `out int value` C# 7.x feature
<!--
You may drag & drop the attachment (repro code/solution, screenshot, etc.) onto the issue.
-->

### Current Behavior

```
[ERROR] FATAL UNHANDLED EXCEPTION: Mono.CSharp.InternalErrorException: Program.cs(5,28): ConsoleApplication16.Program.Main() ---> Mono.CSharp.InternalErrorException: Program.cs(6,9): ---> System.NullReferenceException: Object reference not set to an instance of an object
  at IKVM.Reflection.Emit.ILGenerator.Emit (IKVM.Reflection.Emit.OpCode opc, IKVM.Reflection.Emit.LocalBuilder local) [0x0002a] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.EmitContext.Emit (IKVM.Reflection.Emit.OpCode opcode, IKVM.Reflection.Emit.LocalBuilder local) [0x00000] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.LocalVariable.Emit (Mono.CSharp.EmitContext ec) [0x00011] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.VariableReference.EmitLoad (Mono.CSharp.EmitContext ec) [0x00006] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.VariableReference.Emit (Mono.CSharp.EmitContext ec, System.Boolean leave_copy) [0x00014] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.UnaryMutator.Emit (Mono.CSharp.EmitContext ec) [0x00008] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.VariableReference.EmitAssign (Mono.CSharp.EmitContext ec, Mono.CSharp.Expression source, System.Boolean leave_copy, System.Boolean prepare_for_load) [0x00070] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.UnaryMutator.EmitCode (Mono.CSharp.EmitContext ec, System.Boolean is_expr) [0x0000e] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.UnaryMutator.EmitStatement (Mono.CSharp.EmitContext ec) [0x00000] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.StatementExpression.DoEmit (Mono.CSharp.EmitContext ec) [0x00000] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.Statement.Emit (Mono.CSharp.EmitContext ec) [0x0000d] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.While.DoEmit (Mono.CSharp.EmitContext ec) [0x000cf] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.Statement.Emit (Mono.CSharp.EmitContext ec) [0x0000d] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.Block.DoEmit (Mono.CSharp.EmitContext ec) [0x00010] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.ExplicitBlock.Emit (Mono.CSharp.EmitContext ec) [0x00042] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.ParametersBlock.Emit (Mono.CSharp.EmitContext ec) [0x00030] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.ToplevelBlock.Emit (Mono.CSharp.EmitContext ec) [0x00038] in <5179cae3de0849a38269d13aeda22ca8>:0 
   --- End of inner exception stack trace ---
  at Mono.CSharp.ToplevelBlock.Emit (Mono.CSharp.EmitContext ec) [0x000bb] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.MethodData.Emit (Mono.CSharp.TypeDefinition parent) [0x0008c] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.MethodOrOperator.Emit () [0x001e6] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.Method.Emit () [0x00104] in <5179cae3de0849a38269d13aeda22ca8>:0 
   --- End of inner exception stack trace ---
  at Mono.CSharp.Method.Emit () [0x00114] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.TypeDefinition.Emit () [0x0027b] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.ClassOrStruct.Emit () [0x0001d] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.Class.Emit () [0x00000] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.TypeDefinition.EmitContainer () [0x0000c] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.TypeContainer.EmitContainer () [0x00018] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.NamespaceContainer.EmitContainer () [0x00007] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.TypeContainer.EmitContainer () [0x00018] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.NamespaceContainer.EmitContainer () [0x00007] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.TypeContainer.EmitContainer () [0x00018] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.ModuleContainer.EmitContainer () [0x00085] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.AssemblyDefinition.Emit () [0x0008c] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.AssemblyDefinitionStatic.Emit () [0x0010e] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.Driver.Compile () [0x002a8] in <5179cae3de0849a38269d13aeda22ca8>:0 
  at Mono.CSharp.Driver.Main (System.String[] args) [0x0004c] in <5179cae3de0849a38269d13aeda22ca8>:0 
e
```

### Expected Behavior

csc (Roslyn) works fine.

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:

5.17 (master)
<!--
You can use `mono --version` or About dialog to obtain this information.
-->

--- a/mcs/mcs/expression.cs
+++ b/mcs/mcs/expression.cs
@@ -5679,6 +5679,12 @@ namespace Mono.CSharp
 				ConvCast.Emit (ec, enum_conversion);
 		}
 
+		public override void EmitPrepare (EmitContext ec)
+		{
+			Left.EmitPrepare (ec);
+			Right.EmitPrepare (ec);
+		}
+
 		public override void EmitSideEffect (EmitContext ec)
 		{
 			if ((oper & Operator.LogicalMask) != 0 ||

--- a/mcs/tests/test-decl-expr-05.cs
+++ b/mcs/tests/test-decl-expr-05.cs
@@ -6,6 +6,11 @@ class X
 		{
 			arg = s.ToString ();
 		}
+
+		while (true && Call (out string s2))
+		{
+			arg = s2.ToString ();
+		}
 	}
 
 	static bool Call (out string s)

--- a/mcs/tests/ver-il-net_4_x.xml
+++ b/mcs/tests/ver-il-net_4_x.xml
@@ -68778,7 +68778,7 @@
   <test name="test-decl-expr-05.cs">
     <type name="X">
       <method name="Void Test(System.String)" attrs="129">
-        <size>29</size>
+        <size>62</size>
       </method>
       <method name="Boolean Call(System.String ByRef)" attrs="145">
         <size>17</size>

